dev/bots/custom_rules/debug_assert.dart

```diff
@@ -0,0 +1,382 @@
+  _DebugAssertVisitor(this.unit, this.overridableMemberLookup);
+
+  final List<_DebugOnlySymbolAccessError> accessViolations = <_DebugOnlySymbolAccessError>[];
+  final List<_IncorrectAnnotationError> incorrectAnnotations = <_IncorrectAnnotationError>[];
+
+  final ResolvedUnitResult unit;
+  final Map<ExecutableElement, bool> overridableMemberLookup;
+
+  static bool _isDebugAssertAnnotationElement(ElementAnnotation? annotation) {
+    final Element? annotationElement = annotation?.element;
+    return annotationElement is PropertyAccessorElement && annotationElement.name == 'debugAssert';
+  }
+
+  bool _overriddableClassMemberHasDebugAnnotation(ExecutableElement classMember, InterfaceElement enclosingElement) {
+    assert(!enclosingElement.library.isInSdk);
+    final bool? cached = overridableMemberLookup[classMember];
+    if (cached != null) {
+      return cached;
+    }
+    final bool isClassMemberAnnotated = enclosingElement.library.metadata.any(_isDebugAssertAnnotationElement)
+                                     || classMember.metadata.any(_isDebugAssertAnnotationElement)
+                                     || classMember.enclosingElement.metadata.any(_isDebugAssertAnnotationElement);
+
+    bool isAnnotated = isClassMemberAnnotated;
+    for (final InterfaceType supertype in enclosingElement.allSupertypes) {
+      if (supertype.element.library.isInSdk) {
+        continue;
+      }
+      final ExecutableElement? superImpl = switch (classMember) {
+        MethodElement(:final String name)                           => supertype.getMethod(name),
+        PropertyAccessorElement(:final String name, isGetter: true) => supertype.getGetter(name),
+        PropertyAccessorElement(:final String name, isSetter: true) => supertype.getSetter(name),
+        _ => throw StateError('not reachable $classMember(${classMember.runtimeType})'),
+      };
+      if (superImpl == null) {
+        continue;
+      }
+      final bool isSuperImplAnnotated = _overriddableClassMemberHasDebugAnnotation(superImpl, supertype.element);
+      if (isClassMemberAnnotated && !isSuperImplAnnotated) {
+        incorrectAnnotations.add(_IncorrectAnnotationError(superImpl, classMember));
+      }
+      isAnnotated |= isSuperImplAnnotated;
+    }
+
+    return overridableMemberLookup[classMember] = isAnnotated;
+  }
+
+  bool _syntheticConstructorHasDebugAnnotation(ConstructorElement constructorElement) {
+    assert(constructorElement.isSynthetic);
+    assert(constructorElement.isDefaultConstructor);
+    if (constructorElement.library.isInSdk) {
+      return false;
+    }
+
+    assert(!constructorElement.isFactory);
+    assert(constructorElement.isDefaultConstructor);
+
+    // Subclasses may "inherit" default constructors from the superclass. Since
+    // constructors can't be invoked by the class members (unlike methods that
+    // can have "bad annotations"). Defer to the first non-synthetic default
```

Comment:
I think this case will be caught by the code in `visitConstructorDeclaration`, since `B.new` is not debug-only but `A.new` is, and `B.new` invokes `A.new` implicitly.